lib/active_model/serializer/adapter/json_api.rb

```diff
@@ -4,9 +4,31 @@ module ActiveModel
   class Serializer
     class Adapter
       class JsonApi < Adapter
+        @root = :data
+        DEFAULT_ATTRIBUTES = [:type, :id]
+
+        def self.params(permitted, associations)
+          relationships = {}
+          associations.each do |assoc|
+             relationships[assoc] = {}
+             relationships[assoc][@root] = DEFAULT_ATTRIBUTES
+           end
+
+          return :type, {attributes: permitted}, {relationships: relationships}
+        end
+
+        def self.parse(params)
+          attrs, assoc = {}, {}
+          attrs = params['attributes'] if params['attributes']
+          assoc = params['relationships'].map {|a| {a.shift => a.first['data']['type'].camelize.safe_constantize.find(a.shift['data']['id'])}} if params['relationships']
```

Comment:
Also, as per the specification, the `data` attribute of a relationship object can be an array of resource identifiers. Hence, that part should actually be something along the lines of:
```ruby                                                                                                                                                                                                       
        def self.parse(params)
          attrs, assoc = {}, {}
          attrs = params['attributes'] if params['attributes']
          assoc = params['relationships'].map do |rel|                                                                                                                                                      
            key, data = rel.shift, rel.first['data']                                                                                                                                                        
            value = if data.kind_of? Array                                                                                                                                                                  
                      data.map { |ri| self.parse_resource_identifier ri }                                                                                                                                   
                    elsif data                                                                                                                                                                              
                      self.parse_resource_identifier data                                                                                                                                                   
                    else                                                                                                                                                                                    
                      nil                                                                                                                                                                                   
                    end                                                                                                                                                                                     
            {key => value}                                                                                                                                                                                  
          end if params['relationships']                                                                                                                                                                    
          assoc.reduce attrs, :merge
        end

      private:
        def self.parse_resource_identifier(ri)                                                                                                                                                              
          ri['type'].camelize.singularize.safe_constantize.find(ri['id'])                                                                                                                                   
        end         
```